src/libraries/System.Net.Quic/src/System/Net/Quic/NetEventSource.Quic.Counters.cs

```diff
@@ -0,0 +1,121 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Diagnostics.Metrics;
+using System.Net.Quic;
+
+using Microsoft.Quic;
+using static Microsoft.Quic.MsQuic;
+
+namespace System.Net
+{
+    internal sealed partial class NetEventSource
+    {
+        private static Meter s_meter = new Meter("Private.InternalDiagnostics.System.Net.Quic");
+        private static readonly long[] s_counters = new long[(int)QUIC_PERFORMANCE_COUNTERS.MAX];
+
+        public static readonly ObservableGauge<long> MsQuicCountersGauge = s_meter.CreateObservableGauge<long>(
+            name: "MsQuic",
+            observeValues: GetGauges,
+            unit: null,
+            description: "MsQuic performance counters");
+
+        public static readonly ObservableCounter<long> MsQuicCountersCounter = s_meter.CreateObservableCounter<long>(
+            name: "MsQuic",
+            observeValues: GetCounters,
+            unit: null,
+            description: "MsQuic performance counters");
+
+        [NonEvent]
+        private static void UpdateCounters()
```

Comment:
This is called from both `GetGauges` and `GetCounters`. Is it possible that these will be called concurrently? What will happen? Could it crash the process?

Reply:
It would be fine the two writes would just write same/similar data but should not cause confusion overall.